example/bwmf/bwmf_task.go

```diff
@@ -337,83 +325,52 @@ func (t *bwmfTask) doDataReady(ctx context.Context, fromID uint64, method string
 		if t.epoch%2 == 0 {
 			t.logger.Printf("Full D ready, task %d, epoch %d", t.taskID, t.epoch)
 			// XXX Starting an intensive computation.
-			go t.updateTShard(ctx)
+			if t.dims.M == 0 {
+				t.dims.M = 0
+				for _, m := range t.peerShards {
+					t.dims.M += m.M
+				}
+			}
+			go t.updateShard(ctx, t.dims.M, t.dims.n, t.dims.k, t.tParam, t.tShard, t.columnShard)
 		} else {
 			t.logger.Printf("Full T ready, task %d, epoch %d", t.taskID, t.epoch)
 			// XXX Starting an intensive computation.
-			go t.updateDShard(ctx)
+			if t.dims.N == 0 {
+				t.dims.N = 0
+				for _, m := range t.peerShards {
+					t.dims.N += m.M
+				}
+			}
+			go t.updateShard(ctx, t.dims.N, t.dims.m, t.dims.k, t.dParam, t.dShard, t.rowShard)
 		}
 	}
 }
 
-// These two function carry out actual optimization.
-func (t *bwmfTask) updateDShard(ctx context.Context) {
-	if t.dims.N == -1 {
-		t.dims.N = 0
-		for _, m := range t.peerShards {
-			t.dims.N += len(m.Row)
-		}
-	}
-
+func (t *bwmfTask) updateShard(ctx context.Context, m, n, k uint32,
+	param op.Parameter, shard *pb.MatrixShard, full *pb.MatrixShard) {
```

Comment:
The last two parameters should be: `shard, full *pb.MatrixShard`. Also, `full` is actually the "origin" matrix shard. It's somehow confusing since the collection of peers shards are also mentioned as "fullD" and "fullT".